src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -584,34 +584,41 @@ private void ProcessDownloadedBlock(ChainedHeaderBlock chainedHeaderBlock)
-        private async Task<ConnectBlocksResult> ConnectNewChainAsync(ChainedHeader newTip, ChainedHeader currentTip, List<ChainedHeaderBlock> blocksToConnect)
+        private async Task<ConnectBlocksResult> ConnectChainAsync(ChainedHeader newTip, ChainedHeader currentTip, List<ChainedHeaderBlock> blocksToConnect)
         {
             this.logger.LogTrace("({0}:'{1}',{2}:'{3}',{4}.{5}:{6})", nameof(newTip), newTip, nameof(currentTip), currentTip, nameof(blocksToConnect), nameof(blocksToConnect.Count), blocksToConnect.Count);
 
-            ConnectBlocksResult connectBlockResult = await this.ConnectBlocksAsync(currentTip, blocksToConnect).ConfigureAwait(false);
+            ChainedHeader lastValidatedBlockHeader = null;
+            ConnectBlocksResult connectBlockResult = null;
 
-            if (!connectBlockResult.Succeeded)
+            foreach (ChainedHeaderBlock blockToConnect in blocksToConnect)
             {
-                this.logger.LogTrace("(-)[FAILED_TO_CONNECT]:'{0}'", connectBlockResult);
-                return connectBlockResult;
-            }
+                connectBlockResult = await this.ConnectBlockAsync(currentTip, blockToConnect).ConfigureAwait(false);
 
-            // Blocks connected successfully.
-            List<int> peersToResync = this.SetConsensusTip(newTip);
+                if (!connectBlockResult.Succeeded)
+                {
+                    connectBlockResult.LastValidatedBlockHeader = lastValidatedBlockHeader;
 
-            await this.ResyncPeersAsync(peersToResync).ConfigureAwait(false);
+                    this.logger.LogTrace("(-)[FAILED_TO_CONNECT]:'{0}'", connectBlockResult);
+                    return connectBlockResult;
+                }
 
-            if (this.network.Consensus.MaxReorgLength != 0)
-            {
-                int newFinalizedHeight = newTip.Height - (int)this.network.Consensus.MaxReorgLength;
+                lastValidatedBlockHeader = blockToConnect.ChainedHeader;
 
-                await this.finalizedBlockHeight.SaveFinalizedBlockHeightAsync(newFinalizedHeight).ConfigureAwait(false);
-            }
+                // Block connected successfully.
+                List<int> peersToResync = this.SetConsensusTip(newTip);
+
+                await this.ResyncPeersAsync(peersToResync).ConfigureAwait(false);
+
+                if (this.network.Consensus.MaxReorgLength != 0)
+                {
+                    int newFinalizedHeight = newTip.Height - (int)this.network.Consensus.MaxReorgLength;
+
+                    await this.finalizedBlockHeight.SaveFinalizedBlockHeightAsync(newFinalizedHeight).ConfigureAwait(false);
+                }
 
-            foreach (ChainedHeaderBlock validatedBlock in blocksToConnect)
-            {
                 // TODO: change signal to take ChainedHeaderBlock
-                this.signals.SignalBlock(validatedBlock.Block);
+                this.signals.SignalBlockConnected(blockToConnect.Block);
```

Comment:
This is a bit of a change from before, it means we will be signalling blocks while CM is in a special state, did we consider all the consequences?

Reply:
Yes, as long as callback methods don't call the CM itself in a way that it will try to take reorg lock it should be totally fine.